File: trip-server/scripts/benchmark/benchmark-http.ts
/**
 * 场景 1：普通 HTTP 压测
 *
 * 用 autocannon 测登录 / 查历史 接口
 * 输出 QPS / P50 / P95 / P99 / 错误率
 */

// @ts-ignore — autocannon 没有官方 .d.ts，用 require interop
import autocannon from 'autocannon'
import { saveResult, getEnv } from './lib/result-store'
import { getAuthToken, getEvalCredentials } from './lib/auth'

const BASE_URL = process.env.BASE_URL || 'http://localhost:3000'
const DURATION = 30  // seconds

async function main() {
  console.log('[http] 启动普通 HTTP 压测...')

  // 0) 先拿 token（避免 autocannon 登录压测触发限流后取不到 token）
  const token = await getAuthToken(BASE_URL)
  console.log(`[http] 拿到 token，长度=${token.length}`)

  // 1) 登录接口压测
  const loginResult = await autocannon({
    url: `${BASE_URL}/api/user/login`,
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(getEvalCredentials()),
    connections: 10,
    duration: DURATION,
  })

  // 3) 查历史接口
  const historyResult = await autocannon({
    url: `${BASE_URL}/api/history/trips`,
    method: 'GET',
    headers: { Authorization: `Bearer ${token}` },
    connections: 10,
    duration: DURATION,
  })

  // 合并结果
  const loginSuccess = loginResult.requests.total - loginResult.non2xx - loginResult.errors
  const historySuccess = historyResult.requests.total - historyResult.non2xx - historyResult.errors

  const result = {
    scenario: 'http',
    env: getEnv(),
    config: {
      connections: 10,
      durationSec: DURATION,
      rateLimit: '20 req/min per (IP for login / userId for authed)',
    },
    login: {
      url: `${BASE_URL}/api/user/login`,
      qps: loginResult.requests.average,
      effectiveQps: +(loginSuccess / DURATION).toFixed(3),
      p50: loginResult.latency.p50,
      p95: loginResult.latency.p97_5,
      p99: loginResult.latency.p99,
      max: loginResult.latency.max,
      totalRequests: loginResult.requests.total,
      success2xx: loginSuccess,
      errors: loginResult.errors,
      non2xx: loginResult.non2xx,
      successRate: +(loginSuccess / loginResult.requests.total).toFixed(6),
    },
    history: {
      url: `${BASE_URL}/api/history/trips`,
      qps: historyResult.requests.average,
      effectiveQps: +(historySuccess / DURATION).toFixed(3),
      p50: historyResult.latency.p50,
      p95: historyResult.latency.p97_5,
      p99: historyResult.latency.p99,
      max: historyResult.latency.max,
      totalRequests: historyResult.requests.total,
      success2xx: historySuccess,
      errors: historyResult.errors,
      non2xx: historyResult.non2xx,
      successRate: +(historySuccess / historyResult.requests.total).toFixed(6),
    },
    notes: [
      '10 connections × 30s 触发服务端 rate limit（max=20/min/user）',
      'qps = autocannon 报告的请求率（包含被 429 拒的）',
      'effectiveQps = 2xx / duration，是接口真实吞吐',
      'p50/p95/p99 在被 429 大量污染时意义有限（max 来自少量成功样本）',
    ].join('; '),
  }
  saveResult('http-results', result)
  console.log('[http] 完成')
  console.log(`  登录 raw QPS: ${result.login.qps.toFixed(1)}, effectiveQps: ${result.login.effectiveQps}, P99: ${result.login.p99}ms, 2xx: ${result.login.success2xx}/${result.login.totalRequests}`)
  console.log(`  历史 raw QPS: ${result.history.qps.toFixed(1)}, effectiveQps: ${result.history.effectiveQps}, P99: ${result.history.p99}ms, 2xx: ${result.history.success2xx}/${result.history.totalRequests}`)
}

main().catch((e) => { console.error(e); process.exit(1) })

File: trip-server-py/scripts/benchmark_http.py
"""HTTP 接口压测脚本（对应 Node 版 benchmark-http.ts）

测试端点：
  - POST /api/user/login    登录接口
  - GET  /api/history/trips  历史记录接口（需认证）

使用 asyncio + httpx 实现并发压测，统计 QPS、P50/P95/P99 延迟。
"""

import asyncio
import os
import time

import httpx

from benchmark_lib.auth import get_auth_header, get_eval_credentials, BASE_URL
from benchmark_lib.store import get_env, save_result

CONNECTIONS = int(os.getenv("BENCH_CONNECTIONS", "10"))
DURATION_SEC = int(os.getenv("BENCH_DURATION", "30"))


async def bench_endpoint(client: httpx.AsyncClient, method: str, url: str,
                         body: dict | None = None,
                         headers: dict | None = None) -> list[dict]:
    """对单个端点并发压测"""
    results = []
    start_time = time.monotonic()
    tasks = []

    async def _request():
        nonlocal results
        req_start = time.monotonic()
        try:
            if method == "POST":
                resp = await client.post(url, json=body, headers=headers)
            else:
                resp = await client.get(url, headers=headers)
            elapsed = (time.monotonic() - req_start) * 1000  # ms
            results.append({
                "status": resp.status_code,
                "elapsed_ms": elapsed,
                "success": resp.is_success,
            })
        except Exception as e:
            elapsed = (time.monotonic() - req_start) * 1000
            results.append({"status": 0, "elapsed_ms": elapsed, "success": False, "error": str(e)})

    while time.monotonic() - start_time < DURATION_SEC:
        tasks.append(_request())
        if len(tasks) >= CONNECTIONS:
            await asyncio.gather(*tasks)
            tasks = []

    if tasks:
        await asyncio.gather(*tasks)

    return results


async def main():
    print(f"HTTP Benchmark: {CONNECTIONS} connections × {DURATION_SEC}s")
    print(f"Base URL: {BASE_URL}")

    # 获取认证 token
    print("Getting auth token...")
    auth_headers = get_auth_header(BASE_URL)
    login_body = get_eval_credentials()

    async with httpx.AsyncClient(base_url=BASE_URL, timeout=30) as client:
        # 登录压测
        print(f"Benchmarking POST /api/user/login ...")
        login_results = await bench_endpoint(
            client, "POST", "/api/user/login", body=login_body
        )

        # 历史记录压测
        print(f"Benchmarking GET /api/history/trips ...")
        history_results = await bench_endpoint(
            client, "GET", "/api/history/trips", headers=auth_headers
        )

    # 统计结果
    def stats(results: list[dict], label: str) -> dict:
        total = len(results)
        success = sum(1 for r in results if r["status"] == 200)
        non2xx = sum(1 for r in results if r["status"] != 0 and r["status"] != 200)
        errors = sum(1 for r in results if r["status"] == 0)
        latencies = [r["elapsed_ms"] for r in results if r["status"] == 200]
        all_latencies = [r["elapsed_ms"] for r in results]

        qps = total / DURATION_SEC
        effective_qps = success / DURATION_SEC if success else 0

        sorted_lat = sorted(latencies) if latencies else [0]

        def pct(p):
            if not sorted_lat:
                return 0
            idx = int((len(sorted_lat) - 1) * p / 100)
            return round(sorted_lat[idx], 2)

        result = {
            "url": f"{BASE_URL.rstrip('/')}/{label.lower().replace(' ', '/').replace('//', '/')}",
            "qps": round(qps, 2),
            "effectiveQps": round(effective_qps, 3),
            "p50": pct(50),
            "p95": pct(95),
            "p99": pct(99),
            "max": round(max(all_latencies), 2) if all_latencies else 0,
            "totalRequests": total,
            "success2xx": success,
            "errors": errors,
            "non2xx": non2xx,
            "successRate": round(success / total, 6) if total else 0,
        }
        print(f"  {label}: QPS={result['effectiveQps']}, P99={result['p99']}ms, "
              f"success={success}/{total}")
        return result

    login_stats = stats(login_results, "login")
    history_stats = stats(history_results, "history")

    result = {
        "scenario": "http",
        "env": get_env(),
        "config": {
            "connections": CONNECTIONS,
            "durationSec": DURATION_SEC,
            "rateLimit": "20 req/min per (IP for login / userId for authed)",
        },
        "login": login_stats,
        "history": history_stats,
        "notes": f"{CONNECTIONS} connections × {DURATION_SEC}s. "
                 "Login triggers rate limit (max=20/min/user). "
                 "qps = totalRequests/duration; effectiveQps = 2xx/duration.",
    }

    save_result("http-results", result)
    print("HTTP benchmark complete.")


if __name__ == "__main__":
    asyncio.run(main())
